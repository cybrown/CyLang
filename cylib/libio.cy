void extern putchar(int c);
void extern putbyte(byte c);
void extern putshort(short c);
void extern putint(int c);
void extern putlong(long c);
void extern putfloat(float f);
void extern putdouble(double d);
void extern puts(byte[] str);
void extern printf(byte[] str);

void print(byte c) {
	putbyte(c);
}

void print(short c) {
	putshort(c);
}

void print(int c) {
	putint(c);
}

void print(long c) {
	putlong(c);
}

void print(float f) {
	putfloat(f);
}

void print(double d) {
	putdouble(d);
}

void print(byte[] str) {
	printf(str);
}

void print(bool c) {
	print(c ? "true" : "false");
}
