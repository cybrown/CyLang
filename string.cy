class String {

	byte[] data;
	long _length;

	byte[] getData() {
		byte[] res;
		res = {byte * this._length};
		long i;
		for (i = 0l; i < this._length; i++)
		{
			res[i] = this.data[i];
		}
		return res;
	}

	void setData(byte[] data, long size) {
		long i;
		this.data = {byte * size};
		for (i = 0l; i < size; i++)
		{
			this.data[i] = data[i];
		}
		this._length = size;
	}

	void setData(byte[] data) {
		long i;
		for (i = 0l; data[i] != 0b; i++) {}
		this.setData(data, i);
	}

	long length() {
		return this._length;
	}

	static String create(byte[] data) {
		String s = String();
		s.setData(data);
		return s;
	}

	String sub(long begin, long end) {
		long tmp;
		if (begin > end)
		{
			tmp = begin;
			begin = end;
			end = tmp;
		}
		if (begin > this._length)
			begin = this._length;
		if (end > this._length)
			end = this._length;
		long i;
		byte[] resb;
		resb = {byte * end-begin +1l};
		for (i = begin; i < end; i++)
		{
			resb[i-begin] = this.data[i];
		}
		resb[i] = 0b;
		return String.create(resb);
	}

	String concat(String other) {
		byte[] res;
		res = {byte * this.length() + other.length() -1l};
		long i;
		for (i = 0l; i < this.length(); i++)
		{
			res[i] = this.data[i];
		}
		long cur;
		cur = i;
		for (i = 0l; i < other.length(); i++)
		{
			res[i+cur] = other.at(i);
		}
		res[this.length()+other.length()-1l] = 0b;
		return String.create(res);
	}

	String sub(long begin) {
		return this.sub(begin, this._length);
	}

	byte at(long pos) {
		return this.data[pos];
	}
}
