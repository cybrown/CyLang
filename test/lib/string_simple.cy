import libio;
import string;

byte[] data1 = "Hello, ";
byte[] data2 = "World !";

void print(String s) {
	print(s.getData());
}

String s1 = String.create(data1);
String s2 = String.create(data2);

String s = s1.concat(s2);

print(data1); print("\n");
print(s);; print("\n");
data1[4] = '@';
print(data1);; print("\n");
print(s);; print("\n");
data1 = s.getData();
print(data1);; print("\n");
print(s);; print("\n");
print(s.sub(3l, 5l));; print("\n");
print(s.sub(4l));; print("\n");

print(String.create("Hello, ").concat(String.create("World !")));
print("\n");
