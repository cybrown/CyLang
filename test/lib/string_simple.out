Hello, 
Hello, World 
Hell@, 
Hello, World 
Hello, World 
Hello, World 
lo
o, World 
Hello, World 
